pkg/cmd/kind/test/internal/testcmd/testcmd.go

```diff
@@ -0,0 +1,372 @@
+			}
+
+			// actually run the command
+			return runGinkgo(logger, flags, details)
+		},
+	}
+	cmd.Flags().StringVar(&flags.IPFamily, "ip-family", string(v1alpha4.IPv4Family), "test cluster IP family")
+	cmd.Flags().StringVar(&flags.Name, "name", details.Name, "test cluster context name")
+	cmd.Flags().StringVar(&flags.Artifacts, "artifacts-dir", "", "the directory in which to output test results")
+	cmd.Flags().StringVar(&flags.ImageName, "image", "kindest/node:kubernetes-presubmit", "name:tag of the built kind node image")
+	return cmd
+}
+
+// TODO: junit output for steps a la kubetest
+func runGinkgo(logger log.Logger, flags *flagpole, details Details) (err error) {
+	// first debug the version we're using
+	logger.V(0).Infof("Testing with %s", version.DisplayVersion())
+
+	// then build
+	if err := buildWithGinkgoAndTests(logger, flags); err != nil {
+		return err
+	}
+
+	// setup shared provider for doing cluster things
+	provider := cluster.NewProvider(
+		cluster.ProviderWithLogger(logger),
+	)
+
+	// Check if the cluster already exists
+	if err := checkAlreadyExists(provider, flags.Name); err != nil {
+		return err
+	}
+
+	// then before we do anything more, arrange for cleanup to happen
+	cleanup := func(calledFromSignalHandler bool) {
+		err2 := provider.CollectLogs(flags.Name, filepath.Join(flags.Artifacts, "logs"))
+		err3 := provider.Delete(flags.Name, flags.kubeconfigPath())
+		// don't touch err if we're being called from another goroutine
+		// on program early exit, or if err is already non-nil
+		if calledFromSignalHandler || err != nil {
+			return
+		} else if err2 != nil {
+			err = err2
+		} else {
+			err = err3
+		}
+	}
+	var cleanupOnce sync.Once
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, syscall.SIGTERM, os.Interrupt)
+	go func() {
+		<-signals
+		cleanupOnce.Do(func() { cleanup(true) })
+		os.Exit(1)
+	}()
+	defer func() {
+		cleanupOnce.Do(func() { cleanup(false) })
+	}()
+
+	// test (including cluster bringup)
+	return testGinkgo(logger, provider, flags, details)
+}
+
+func buildWithGinkgoAndTests(logger log.Logger, flags *flagpole) error {
+	// primarily for CI we allow overriding the build mode
+	buildType := os.Getenv("KIND_BUILD_TYPE")
+	switch buildType {
+	case "bazel", "docker":
+	default:
+		buildType = "docker"
+	}
+
+	// possibly enable bazel build caching before building kubernetes
+	// TODO: remove this if we move to RBE ..?
+	if os.Getenv("BAZEL_REMOTE_CACHE_ENABLED") == "true" {
+		_ = exec.Command("create_bazel_cache_rcs.sh").Run()
+	}
+
+	// first build the node image
+	ctx, err := node.NewBuildContext(
+		node.WithMode(buildType),
+		node.WithImage(flags.ImageName),
+		node.WithLogger(logger),
+	)
+	if err != nil {
+		return errors.Wrap(err, "error creating build context")
+	}
+	if err := ctx.Build(); err != nil {
+		return errors.Wrap(err, "error building node image")
+	}
+
+	// build test binaries
+	switch buildType {
+	case "bazel":
+		if err := exec.InheritOutput(
```

Comment:
this writes to the logger, right?

Reply:
no. this wires it to stdout / stderr.

this respecting streams is a broad TODO, but it's not super important for cmd/ at the moment